incubator/honeydipper/templates/daemon-deployment.yaml

```diff
@@ -0,0 +1,115 @@
+apiVersion: apps/v1beta2
+kind: Deployment
+metadata:
+  name: {{ include "honeydipper.fullname" . }}
+  labels:
+    app.kubernetes.io/name: {{ include "honeydipper.name" . }}
+    helm.sh/chart: {{ include "honeydipper.chart" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+spec:
+  replicas: {{ .Values.daemon.replicaCount }}
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ include "honeydipper.name" . }}
+      app.kubernetes.io/instance: {{ .Release.Name }}
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: {{ include "honeydipper.name" . }}
+        app.kubernetes.io/instance: {{ .Release.Name }}
+    spec:
+      {{- with .Values.daemon.spec }}
```

Comment:
I find this a little bit too generic.

Reply:
Regarding the `.Values.daemon.spec`,  I appreciate your input. In our environment, we use this to add configurations such as  `dnsConfig`, `serviceAccountName` etc. It is hard to predict what will be needed down the road, hence a catch all approach is created to deal with this.  Without this, we will have to list all possible items from the kubernetes v1 pod spec in the templates.  I don't want to take away the possibility for my user to have fine-grained control over the deployment. 

An observation on this, which probably deserve a separate thread to discuss: a lot of the available helm charts (atlantis, spinnaker just to name a few) do keep their charts simple by not providing this type of control to the users.  I feel this to be a blocker for the adoption of these charts, and sometimes even for the adoption of helm itself.  Almost every chart we are using in our organization has to be tweaked and just so some tiny configuration option can be made possible. This is one of the reason that the majority of the deployments in my organization is just plain kubernetes manifest files with some other templating solution. I would appreciate if you can point me to a direction to achieve a balance between keeping the options simple, specific and providing fine-grained control.

With all that said, I will change this if you insist.

Reply:
I think it's important to highlight two things. First, Helm is a package manager (think of it like apt for Kubernetes) where charts are the packages. Second, in Helm we've documented the set of role/profiles we target and have put an order to them... https://github.com/helm/community/blob/master/user-profiles.md

Consider the case for someone using apt to install mysql or postgresql on ubuntu or debian. A typical person doing this isn't an ubuntu/debian expert and it's a mysql or postgresql expert. The person who builds the package is able to make the application installation accessible to non-experts.

What I wonder is, who is the target audience for honeydipper and what are we expecting them to know? How does the chart make it accessible to those people?

Reply:
@mattfarina I don't think the comparison between helm and apt is 100% valid.  You don't apt install apache twice to get two web server installations, but you can helm install apache multiple times with different names and values files.  IMHO, Helm definitely does more than just package management.  But, I totally agree and understand that there should be a boundary on what helm should and should not do.  And that is exactly why I am using a generic parameter in the chart values. It is not helm's responsibility to tell `application operators` what option they can or can not use.

I think I fit into the `application developer` and `application distributor` profile. The values file is a contract between me and the `application operators`. I would like to enable the `application operators` to use whatever available options kubernetes provides. Using a generic parameter is a way to achieve that. We are already using `dnsConfig` and `serviceAccount` in the `spec`, and I don't know what else we (or other people) might be using, could be anything in https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#pod-v1-core 

In my case, the intended users for Honeydipper are SREs, system engineers, maybe some software engineers. I think they (I for one) would appreciate to have a better control over the deployment of the tool when using helm charts.

Reply:
@Charles546 ...

> I don't think the comparison between helm and apt is 100% valid. You don't apt install apache twice to get two web server installations, but you can helm install apache multiple times with different names and values files. IMHO, Helm definitely does more than just package management.

Kubernetes and Debian/Ubuntu are different operating systems. These GNU/Linux distros are for a single server. Kubernetes was designed as a cluster operating system and comes from the idea of (warehouse sized computers at Google](https://ai.google/research/pubs/pub35290). The number of instances of an application you run on a single server is different from that of a large scale computer, even those much smaller than a warehouse.

The concept of a package manager still holds true. The differences are rooted in the differences in the operating systems.

> I think I fit into the `application developer` and `application distributor` profile. The values file is a contract between me and the `application operators`.

What do you expect the `application operators` will know about honeydipper and Kubernetes? Why do you think they will know that? For example, how will they know the details of honeydipper well enough to know what to do with the daemonset? What if they don't know the technical details of either well?

While I don't know Honeydipper I do know Kubernetes. K8s is complicated and hard. It's hard to onboard new people to it. A lot of people who claim to know k8s are still not using things like PodDisruptionBudget. I'm finding a lot of people don't even really know it. This is just one example.

How will they know the technical details of Honeydipper with enough knowledge to do wide variation to hand craft their DaemonSet?

I'm asking a lot of UX stuff because a package manager is about UX. Someone can grab mysql, put it on a Linux distro, craft all the config themselves, and put it in the right place themselves. I've done it and I've seen people do that. Especially when they are experts on the stack. Most people use a package manager and don't spend the time trying to learn that knowledge. Package managers let you encapsulate knowledge and automate handing it off to make the experience easier.

Reply:
@mattfarina I get it, we have to accept the difference between kubernetes and ubuntu/debian and at the same time accept the similarity between helm and apt.  I DO accept them both.  But we can not deny that there are differences between helm and apt too, which is the point in my previous comments: helm does more than package management. 

If we accept the fact that helm does also handle configurations, we will be able to go beyond the UX requirement for a package manager and dive into the usability requirements. In fact, introducing the generic `spec` parameter does not affect the UX at all, other than one more line in the documents. We don't force them to use this `spec`, but if this is missing, the `operators` will have no knobs to turn when they need it - that is `lack of usability`. I am anticipating that most of the beginners of Honeydipper won't need to use or care about the `spec`, but as their needs evolve, they will need to customize the deployment one way or another, sooner or later. 

To answer your question, I do expect the users or `application operators` of Honeydipper to know kubernetes pretty well. If they want to use `PodDisruptionBudget` through `spec`, they can and should.  We don't want to take away this knob because some of the users don't understand it at this moment.

The difference between apt and helm is that apt does not handle configuration, or in other words it is transparent to configurations, while helm is not transparent, not totally. The generic `spec` can make helm a little bit more transparent and that is what Honeydipper needs. I hope this explains well my perspective.

Reply:
We are not going to accept the `spec` parameter as is.

Reply:
Changed